src/core/components/pin-set.js

```diff
@@ -0,0 +1,267 @@
+            .sort((a, b) => Buffer.compare(a.link.multihash, b.link.multihash))
+
+          const rootLinks = fanoutLinks.concat(nodes.map(item => item.link))
+          const rootData = Buffer.concat(
+            [headerBuf].concat(nodes.map(item => item.data))
+          )
+
+          DAGNode.create(rootData, rootLinks, (err, rootNode) => {
+            if (err) { return cb(err) }
+            return cb(null, rootNode)
+          })
+        } else {
+          // If the array of pins is > maxItems, we:
+          //  - distribute the pins among `defaultFanout` bins
+          //    - create a DAGNode for each bin
+          //      - add each pin of that bin as a DAGLink
+          //  - create a root DAGNode
+          //    - store each bin as a DAGLink
+          //  - send that root DAGNode via `cb`
+          // (using go-ipfs' "wasteful but simple" approach for consistency)
+
+          const bins = pins.reduce((bins, pin) => {
+            const n = hash(depth, pin.key) % defaultFanout
+            bins[n] = n in bins ? bins[n].concat([pin]) : [pin]
+            return bins
+          }, {})
+
+          const binKeys = Object.keys(bins)
+          binsToFill += binKeys.length
+          binKeys.forEach(n => {
+            storePins(
+              bins[n],
+              (err, child) => storeChild(err, child, n),
+              depth + 1,
+              binsToFill,
+              binsFilled
+            )
+          })
+        }
+
+        function storeChild (err, child, bin) {
+          if (err) { return cb(err) }
+
+          const cid = new CID(child._multihash)
+          dag.put(child, { cid }, (err) => {
+            if (err) { return cb(err) }
+
+            fanoutLinks[bin] = new DAGLink('', child.size, child.multihash)
+            binsFilled++
+
+            if (binsFilled === binsToFill) {
+              // all finished
+              DAGNode.create(headerBuf, fanoutLinks, (err, rootNode) => {
+                if (err) { return cb(err) }
+                return cb(null, rootNode)
+              })
+            }
+          })
+        }
+      }
+    },
+
+    loadSet: (rootNode, name, callback) => {
+      callback = once(callback)
+      const link = rootNode.links.find(l => l.name === name)
+      if (!link) {
+        return callback(new Error('No link found with name ' + name))
+      }
+
+      dag.get(link.multihash, (err, res) => {
+        if (err) { return callback(err) }
+        const keys = []
+        const step = link => keys.push(link.multihash)
+        pinSet.walkItems(res.value, step, err => {
+          if (err) { return callback(err) }
+          return callback(null, keys)
+        })
+      })
+    },
+
+    walkItems: (node, step, callback) => {
+      callback = once(callback)
+      let pbh
+      try {
+        pbh = readHeader(node)
+      } catch (err) {
+        return callback(err)
+      }
+      let subwalkCount = 0
+      let finishedCount = 0
```

Comment:
Again, instead of using counts to determine when async tasks have been finished I'd defer to the async library.